### PR TITLE
PUBDEV-8712: Expose Sequential Walker

### PR DESCRIPTION
**Pull request by tomasfryda**
https://h2oai.atlassian.net/browse/PUBDEV-8712

Currently, SequentialWalker is used only for AutoML's exploitation phase - training one model after another in a pre-defined sequence early stopping if the models don't get better.

This PR exposes it, so it is usable from the clients (R/Py) and adds an early stopping option to disable early stopping so we can do an exhaustive search over the provided parameters. Another addition is `max_models` and `max_runtime_secs` parameters to the SequentialWalker.

### CODE REVIEW

**Review by sebhrusen — APPROVED** [review]
looks great, thanks.
Didn't you mention that we would also need to accept hyperparameters passed as R dataframe (or H2O frame, or pandas?) for this sequential search mode?

File: h2o-core/src/main/java/hex/grid/HyperSpaceSearchCriteria.java [hunk 1]
```diff
@@ -172,16 +178,25 @@ private StoppingCriteria cloneStoppingCriteria() {
   public static final class SequentialSearchCriteria extends HyperSpaceSearchCriteria {
 
     private StoppingCriteria _stoppingCriteria;
+    private boolean _early_stopping; // has to be snake_case to match the field in the schema
```
[the remark this answers is not in the document]
Comment (tomasfryda) [reply h2o-core/src/main/java/hex/grid/HyperSpaceSearchCriteria.java:181]: I will remove the comment, it was just confusing to me seeing `_stoppingCriteria` and `_early_stopping` together.

**Comment by tomasfryda** [comment]
@sebhrusen Yes, I mentioned the data.frame but it was mainly to make it easier to explain ("row" is easier than "elements with the same index").

In R, all that it takes to convert a data.frame to a list that we use here is `as.list(df)` so it wouldn't be hard to implement but in python it would require supporting pandas. So I think it's better to stick to the "basic" data structures as there is no significant improvement when using data frames and the use of data frame probably doesn't make sense for the rest of the strategies. (Intuitively, data.frame would make me think that hyper-params on the same row have some reason for being on the same row which it wouldn't)

**Review by ledell — APPROVED** [review]
This is perfect, thank you @tomasfryda!

File: h2o-core/src/main/java/hex/schemas/GridSearchSchema.java [hunk 2]
```diff
@@ -112,22 +113,15 @@ private static Map<String, Object[]> paramValuesToArray(Map<String, Object> para
...
-          search_criteria = new HyperSpaceSearchCriteriaV99.RandomDiscreteValueSearchCriteriaV99();
-          if (p.containsKey("max_runtime_secs") && Double.parseDouble((String) p.get("max_runtime_secs"))<0) {
-            throw new H2OIllegalArgumentException("max_runtime_secs must be >= 0 (0 for unlimited time)", strategy);
-          }
-          if (p.containsKey("max_models") && Integer.parseInt((String) p.get("max_models"))<0) {
-            throw new H2OIllegalArgumentException("max_models must be >= 0 (0 for all models)", strategy);
-          }
-        } else {
-          throw new H2OIllegalArgumentException("search_criteria.strategy", strategy);
+        HyperSpaceSearchCriteria.Strategy strategy = HyperSpaceSearchCriteria.Strategy.valueOf((String) p.get("strategy"));
```
Comment (sebhrusen) [inline h2o-core/src/main/java/hex/schemas/GridSearchSchema.java:116]: to answer @ledell 's concern about case sensitivity. Using `EnumUtils.valueOf`instead would solve this.

**Pull request merged**